pkg/scalers/cpu_memory_scaler.go

```diff
@@ -142,7 +210,294 @@ func (s *cpuMemoryScaler) GetMetricSpecForScaling(context.Context) []v2.MetricSp
+		}
+
+		if capacity == 0 {
+			continue
+		}
+
+		utilization := (metricValue * 100) / capacity
+		totalUtilization += utilization
+		podCount++
+	}
+
+	if podCount == 0 {
+		return nil, fmt.Errorf("no running pods found with non-zero capacity")
+	}
+
+	averageUtilization := int32(totalUtilization / int64(podCount))
+	return &averageUtilization, nil
+}
+
+// Helper functions
+func getResourceValue(containerMetrics *v1beta1.ContainerMetrics, metricName string) *resource.Quantity {
+	switch metricName {
+	case cpuMetricName:
+		return containerMetrics.Usage.Cpu()
+	case memoryMetricName:
+		return containerMetrics.Usage.Memory()
+	default:
+		return nil
+	}
+}
+
+func getPodResourceValue(podMetrics *v1beta1.PodMetrics, metricName string) *resource.Quantity {
+	var total resource.Quantity
+	for _, container := range podMetrics.Containers {
+		if value := getResourceValue(&container, metricName); value != nil {
+			total.Add(*value)
+		}
+	}
+	return &total
+}
+
+func getResourceValueInMillis(containerMetrics *v1beta1.ContainerMetrics, metricName string) int64 {
+	switch metricName {
+	case cpuMetricName:
+		return containerMetrics.Usage.Cpu().MilliValue()
+	case memoryMetricName:
+		return containerMetrics.Usage.Memory().Value()
+	default:
+		return 0
+	}
+}
+
+func getPodResourceValueInMillis(podMetrics *v1beta1.PodMetrics, metricName string) int64 {
+	var total int64
+	for _, container := range podMetrics.Containers {
+		total += getResourceValueInMillis(&container, metricName)
+	}
+	return total
+}
+
+func getResourceName(metricName string) corev1.ResourceName {
+	switch metricName {
+	case cpuMetricName:
+		return corev1.ResourceCPU
+	case memoryMetricName:
+		return corev1.ResourceMemory
+	default:
+		return ""
+	}
+}
+
+func getPodResourceCapacity(pod *corev1.Pod, resourceName corev1.ResourceName) int64 {
+	var total int64
+	for _, container := range pod.Spec.Containers {
+		if quantity, ok := container.Resources.Requests[resourceName]; ok {
+			total += quantity.MilliValue()
+		}
+	}
+	return total
+}
+
+func (s *cpuMemoryScaler) getPodList(ctx context.Context) (*corev1.PodList, labels.Selector, error) {
+	var labelSelector labels.Selector
+
+	switch s.metadata.ScaleTargetKind {
+	case "Deployment":
+		deployment := &appsv1.Deployment{}
+		err := s.kubeClient.Get(ctx, types.NamespacedName{Namespace: s.metadata.Namespace, Name: s.metadata.ScaleTargetName}, deployment)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get deployment: %w", err)
+		}
+
+		labelSelector = labels.SelectorFromSet(deployment.Spec.Selector.MatchLabels)
+	case "StatefulSet":
+		statefulSet := &appsv1.StatefulSet{}
+		err := s.kubeClient.Get(ctx, types.NamespacedName{Namespace: s.metadata.Namespace, Name: s.metadata.ScaleTargetName}, statefulSet)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get statefulset: %w", err)
+		}
+
+		labelSelector = labels.SelectorFromSet(statefulSet.Spec.Selector.MatchLabels)
+	default:
+		return nil, nil, nil
```

Comment:
What about other types?

See if this can be reused: https://github.com/kedacore/keda/blob/4eb71492f4dbf17951aa5a28a0940a9692c1cd5e/pkg/scaling/resolver/scale_resolvers.go#L96

Reply:
I added GVK as the means to reference the scale target. However in the `default` case, there is still no way to fetch the `MatchLabels` field.

Reply:
you can try to check if we can create a duck type that would desribe the resource, similar as it is done in the referenced code for podspec or here https://github.com/kedacore/keda/blob/main/apis/keda/v1alpha1/withtriggers_types.go

If we are not able to do that, then we should check supported scaletargets when we are creating the scaler and fail in that case